File: frontend/src/lib/api/tabs.ts
import { requestJson, requestVoid } from "./shared";
import type { TaskTab, TabEdge, WorkflowDefinition } from "@/types";

interface TabsResponse {
  tabs: TaskTab[];
}

interface TabDetailNode {
  id: string;
  node_type: string;
  tab_id: string | null;
  role_name: string | null;
  is_leader: boolean;
  state: string;
  connections: string[];
  name: string | null;
  todos: Array<{ text: string; type: string }>;
  position: {
    x: number;
    y: number;
  } | null;
  config?: Record<string, unknown>;
  inputs?: Array<Record<string, unknown>>;
  outputs?: Array<Record<string, unknown>>;
}

interface TabDetailResponse {
  tab: TaskTab;
  nodes: TabDetailNode[];
  edges: TabEdge[];
}

export async function fetchTabs(signal?: AbortSignal): Promise<TaskTab[]> {
  return requestJson<TabsResponse, TaskTab[]>("/api/tabs", {
    method: "GET",
    signal,
    errorMessage: "Failed to fetch tabs",
    fallback: [],
    map: (data) => data?.tabs ?? [],
  });
}

export async function createTabRequest(
  title: string,
  allow_network = false,
  write_dirs: string[] = [],
): Promise<TaskTab> {
  return requestJson<TaskTab, TaskTab>("/api/tabs", {
    method: "POST",
    body: { title, allow_network, write_dirs },
    errorMessage: "Failed to create tab",
  });
}

export async function duplicateTabRequest(tabId: string): Promise<TaskTab> {
  return requestJson<TaskTab, TaskTab>(`/api/tabs/${tabId}/duplicate`, {
    method: "POST",
    errorMessage: "Failed to duplicate workflow",
  });
}

export async function updateTabDefinitionRequest(
  tabId: string,
  definition: WorkflowDefinition,
): Promise<TaskTab> {
  return requestJson<TaskTab, TaskTab>(`/api/tabs/${tabId}/definition`, {
    method: "PUT",
    body: { definition },
    errorMessage: "Failed to update workflow definition",
  });
}

export async function deleteTabRequest(tabId: string): Promise<void> {
  await requestVoid(`/api/tabs/${tabId}`, {
    method: "DELETE",
    errorMessage: "Failed to delete tab",
  });
}

export async function fetchTabDetail(
  tabId: string,
  signal?: AbortSignal,
): Promise<TabDetailResponse> {
  return requestJson<TabDetailResponse, TabDetailResponse>(
    `/api/tabs/${tabId}`,
    {
      method: "GET",
      signal,
      errorMessage: "Failed to fetch tab detail",
    },
  );
}

export async function createTabNodeRequest(
  tabId: string,
  body: {
    node_type?: string;
    role_name?: string | null;
    name?: string | null;
    config?: Record<string, unknown>;
    tools?: string[];
    write_dirs?: string[];
    allow_network?: boolean;
  },
): Promise<TabDetailNode> {
  return requestJson<Record<string, unknown>, TabDetailNode>(
    `/api/tabs/${tabId}/nodes`,
    {
      method: "POST",
      body,
      errorMessage: "Failed to create node",
      map: (data) =>
        ({
          id: typeof data?.id === "string" ? data.id : "",
          node_type:
            typeof data?.node_type === "string" ? data.node_type : "agent",
          tab_id: typeof data?.tab_id === "string" ? data.tab_id : null,
          role_name:
            typeof data?.role_name === "string" ? data.role_name : null,
          is_leader: data?.is_leader === true,
          state: typeof data?.state === "string" ? data.state : "idle",
          connections: Array.isArray(data?.connections)
            ? data.connections.filter(
                (item): item is string => typeof item === "string",
              )
            : [],
          name: typeof data?.name === "string" ? data.name : null,
          todos: Array.isArray(data?.todos)
            ? (data.todos as Array<{ text: string; type: string }>)
            : [],
          position:
            data?.position &&
            typeof data.position === "object" &&
            typeof (data.position as { x?: unknown }).x === "number" &&
            typeof (data.position as { y?: unknown }).y === "number"
              ? (data.position as { x: number; y: number })
              : null,
          config:
            data?.config && typeof data.config === "object"
              ? (data.config as Record<string, unknown>)
              : undefined,
          inputs: Array.isArray(data?.inputs)
            ? (data.inputs as Array<Record<string, unknown>>)
            : undefined,
          outputs: Array.isArray(data?.outputs)
            ? (data.outputs as Array<Record<string, unknown>>)
            : undefined,
        }) satisfies TabDetailNode,
    },
  );
}

export async function deleteTabNodeRequest(
  tabId: string,
  nodeId: string,
): Promise<void> {
  await requestVoid(`/api/tabs/${tabId}/nodes/${nodeId}`, {
    method: "DELETE",
    errorMessage: "Failed to delete node",
  });
}

export async function createTabEdgeRequest(
  tabId: string,
  input: {
    fromNodeId: string;
    fromPortKey?: string;
    toNodeId: string;
    toPortKey?: string;
    kind?: "control" | "data" | "event";
  },
): Promise<TabEdge> {
  return requestJson<TabEdge, TabEdge>(`/api/tabs/${tabId}/edges`, {
    method: "POST",
    body: {
      from_node_id: input.fromNodeId,
      from_port_key: input.fromPortKey ?? "out",
      to_node_id: input.toNodeId,
      to_port_key: input.toPortKey ?? "in",
      kind: input.kind ?? "control",
    },
    errorMessage: "Failed to create edge",
  });
}

export async function deleteTabEdgeRequest(
  tabId: string,
  input: {
    edgeId?: string;
    fromNodeId?: string;
    fromPortKey?: string;
    toNodeId?: string;
    toPortKey?: string;
  },
): Promise<void> {
  const params = new URLSearchParams();
  if (input.edgeId) {
    params.set("edge_id", input.edgeId);
  }
  if (input.fromNodeId) {
    params.set("from_node_id", input.fromNodeId);
  }
  if (input.fromPortKey) {
    params.set("from_port_key", input.fromPortKey);
  }
  if (input.toNodeId) {
    params.set("to_node_id", input.toNodeId);
  }
  if (input.toPortKey) {
    params.set("to_port_key", input.toPortKey);
  }
  await requestVoid(`/api/tabs/${tabId}/edges?${params.toString()}`, {
    method: "DELETE",
    errorMessage: "Failed to delete edge",
  });
}

File: backend/src/flowent_api/routes/tabs.py
from __future__ import annotations

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel, ConfigDict

from flowent_api.graph_service import (
    create_agent_node,
    create_edge,
    create_graph_node,
    create_tab,
    delete_agent_node,
    delete_edge,
    delete_tab,
    duplicate_tab,
    list_node_connection_ids,
    list_tab_edges,
    list_workflow_nodes,
    serialize_tab_summary,
    update_tab_definition,
)
from flowent_api.models import AgentState, EdgeKind, WorkflowNodeKind
from flowent_api.registry import registry
from flowent_api.workspace_store import workspace_store

router = APIRouter()


class CreateTabRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    title: str
    allow_network: bool = False
    write_dirs: list[str] = []


class CreateTabNodeRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    node_type: str = WorkflowNodeKind.AGENT.value
    role_name: str | None = None
    name: str | None = None
    config: dict[str, object] = {}
    tools: list[str] = []
    write_dirs: list[str] = []
    allow_network: bool = False


class CreateTabEdgeRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    from_node_id: str
    from_port_key: str = "out"
    to_node_id: str
    to_port_key: str = "in"
    kind: str = EdgeKind.CONTROL.value


class UpdateTabDefinitionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    definition: dict[str, object]


def _serialize_workflow_node(
    *,
    tab_id: str,
    node_id: str,
) -> dict[str, object]:
    tab = workspace_store.get_tab(tab_id)
    if tab is None:
        raise HTTPException(status_code=404, detail="Workflow not found")
    definition = tab.definition.get_node(node_id)
    if definition is None:
        raise HTTPException(status_code=404, detail="Node not found")
    record = workspace_store.get_node_record(node_id)
    live_node = registry.get(node_id)
    config = dict(definition.config)
    role_name = config.get("role_name")
    name = config.get("name")
    position = tab.definition.view.positions.get(node_id)
    todos = (
        [todo.serialize() for todo in live_node.get_todos_snapshot()]
        if live_node is not None
        else [todo.serialize() for todo in (record.todos if record is not None else [])]
    )
    state = (
        live_node.state.value
        if live_node is not None
        else (record.state.value if record is not None else AgentState.IDLE.value)
    )
    return {
        "id": definition.id,
        "node_type": definition.type.value,
        "workflow_id": tab_id,
        "role_name": role_name if isinstance(role_name, str) else None,
        "is_leader": False,
        "state": state,
        "connections": list_node_connection_ids(tab_id=tab_id, node_id=definition.id)
        if definition.type == WorkflowNodeKind.AGENT
        else [],
        "name": name if isinstance(name, str) else None,
        "todos": todos if definition.type == WorkflowNodeKind.AGENT else [],
        "position": position.serialize() if position is not None else None,
        "config": config,
        "inputs": [port.serialize() for port in definition.inputs],
        "outputs": [port.serialize() for port in definition.outputs],
    }


@router.get("/api/workflows")
async def list_workflows() -> dict[str, object]:
    tabs = workspace_store.list_tabs()
    return {"workflows": [serialize_tab_summary(tab) for tab in tabs]}


@router.post("/api/workflows")
async def create_workflow_route(req: CreateTabRequest) -> dict[str, object]:
    title = req.title.strip()
    if not title:
        raise HTTPException(status_code=400, detail="title must not be empty")
    try:
        tab = create_tab(
            title=title,
            allow_network=req.allow_network,
            write_dirs=req.write_dirs,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return serialize_tab_summary(tab)


@router.post("/api/workflows/{tab_id}/duplicate")
async def duplicate_workflow_route(tab_id: str) -> dict[str, object]:
    duplicated, error = duplicate_tab(tab_id=tab_id)
    if error is not None or duplicated is None:
        raise HTTPException(
            status_code=404 if error and error.endswith("not found") else 400,
            detail=error or "Failed to duplicate workflow",
        )
    return serialize_tab_summary(duplicated)


@router.get("/api/workflows/{tab_id}")
async def get_workflow(tab_id: str) -> dict[str, object]:
    tab = workspace_store.get_tab(tab_id)
    if tab is None:
        raise HTTPException(status_code=404, detail="Workflow not found")
    nodes = [
        _serialize_workflow_node(tab_id=tab_id, node_id=node.id)
        for node in list_workflow_nodes(tab_id)
    ]
    edges = [edge.serialize() for edge in list_tab_edges(tab_id)]
    return {
        "workflow": serialize_tab_summary(tab),
        "nodes": nodes,
        "edges": edges,
    }


@router.put("/api/workflows/{tab_id}/definition")
async def update_workflow_definition_route(
    tab_id: str,
    req: UpdateTabDefinitionRequest,
) -> dict[str, object]:
    updated, error = update_tab_definition(
        tab_id=tab_id,
        definition_payload=req.definition,
        actor_id=tab_id,
    )
    if error is not None or updated is None:
        raise HTTPException(
            status_code=400 if error and not error.endswith("not found") else 404,
            detail=error or "Failed to update workflow definition",
        )
    return serialize_tab_summary(updated)


@router.delete("/api/workflows/{tab_id}")
async def delete_workflow_route(tab_id: str) -> dict[str, object]:
    deleted, error = delete_tab(tab_id=tab_id)
    if error is not None or deleted is None:
        status_code = 404 if error and error.endswith("not found") else 400
        raise HTTPException(
            status_code=status_code,
            detail=error or "Failed to delete workflow",
        )
    return deleted


@router.post("/api/workflows/{tab_id}/nodes")
async def create_workflow_node(
    tab_id: str,
    req: CreateTabNodeRequest,
) -> dict[str, object]:
    try:
        node_type = WorkflowNodeKind(req.node_type.strip())
    except ValueError as exc:
        raise HTTPException(status_code=400, detail="Invalid node_type") from exc

    if node_type == WorkflowNodeKind.AGENT:
        if not isinstance(req.role_name, str) or not req.role_name.strip():
            raise HTTPException(status_code=400, detail="role_name is required")
        record, error = create_agent_node(
            role_name=req.role_name,
            tab_id=tab_id,
            name=req.name,
            tools=req.tools,
            write_dirs=req.write_dirs,
            allow_network=req.allow_network,
        )
        if error is not None or record is None:
            raise HTTPException(
                status_code=400, detail=error or "Failed to create node"
            )
        return _serialize_workflow_node(tab_id=tab_id, node_id=record.id)

    node, error = create_graph_node(
        tab_id=tab_id,
        node_type=node_type,
        config={
            **req.config,
            **(
                {"name": req.name}
                if isinstance(req.name, str) and req.name.strip()
                else {}
            ),
        },
        actor_id=tab_id,
    )
    if error is not None or node is None:
        raise HTTPException(status_code=400, detail=error or "Failed to create node")
    return _serialize_workflow_node(tab_id=tab_id, node_id=node.id)


@router.post("/api/workflows/{tab_id}/edges")
async def create_workflow_edge(
    tab_id: str,
    req: CreateTabEdgeRequest,
) -> dict[str, object]:
    tab = workspace_store.get_tab(tab_id)
    if tab is None:
        raise HTTPException(status_code=404, detail="Workflow not found")
    try:
        edge_kind = EdgeKind(req.kind.strip())
    except ValueError as exc:
        raise HTTPException(status_code=400, detail="Invalid edge kind") from exc
    edge, error = create_edge(
        tab_id=tab_id,
        from_node_id=req.from_node_id,
        from_port_key=req.from_port_key,
        to_node_id=req.to_node_id,
        to_port_key=req.to_port_key,
        kind=edge_kind,
    )
    if error is not None or edge is None:
        raise HTTPException(status_code=400, detail=error or "Failed to create edge")
    return edge.serialize()


@router.delete("/api/workflows/{tab_id}/nodes/{node_id}")
async def delete_workflow_node(tab_id: str, node_id: str) -> dict[str, object]:
    deleted, error = delete_agent_node(
        tab_id=tab_id,
        node_id=node_id,
    )
    if error is not None or deleted is None:
        status_code = 404 if error and error.endswith("not found") else 400
        raise HTTPException(
            status_code=status_code, detail=error or "Failed to delete node"
        )
    return deleted


@router.delete("/api/workflows/{tab_id}/edges")
async def delete_workflow_edge(
    tab_id: str,
    edge_id: str | None = None,
    from_node_id: str | None = None,
    to_node_id: str | None = None,
    from_port_key: str | None = None,
    to_port_key: str | None = None,
) -> dict[str, object]:
    deleted, error = delete_edge(
        tab_id=tab_id,
        edge_id=edge_id,
        from_node_id=from_node_id,
        to_node_id=to_node_id,
        from_port_key=from_port_key,
        to_port_key=to_port_key,
    )
    if error is not None or deleted is None:
        status_code = 404 if error and error.endswith("not found") else 400
        raise HTTPException(
            status_code=status_code, detail=error or "Failed to delete edge"
        )
    return deleted
